### ISSUE TITLE
MaxLifeTime configuration impacting dynamic property changes

### ISSUE THREAD

**Issue by deanfundrecs** [brettwooldridge/HikariCP#1350]
#### Environment
```
HikariCP version: 2.7.9
JDK version     : 1.8.0_202
Database        : MySQL
Driver version  : x.x.x
Spring Boot: 2.0.5.RELEASE
```
Hi I have been doing a lot of research into getting Hikari to work with RDS IAM credentials in a multi tenant environment where each tenant has their own database. We are using spring boot and a brief summary of what occurs at startup is as follows. 

A database containing tenant specific information such as JDBC urls is read and per tenant a new data source object is created. This results in a HikariPool per tenant, which has a ScheduledExecutorService supplied to each pool which refreshes the RDS credentials every 14 minutes through the HikariConfigMXBean. This was a recommended several times on other issues when developers were looking to change properties at run time.

The code used to create the data source is as follows:
```
    /**
     * Creates a data source with the given jdbc url and jdbc username
     * the password is generated for RDS access and the connection uses SSL
     *
     * @param tenantId the id of the tenant which this data source will belong to
     * @param jdbcUrl the url of the data source
     * @param jdbcUsername the username to use to access the data source
     * @return the data source
     */
    private HikariDataSource createDataSource(String tenantId, String jdbcUrl, String jdbcUsername){
        HikariConfig config = new HikariConfig()
        config.setDriverClassName(dataSourceProperties.getDriverClassName())
        config.addDataSourceProperty("useSSL", true)
        config.addDataSourceProperty("requireSSL", true)
        config.addDataSourceProperty("verifyServerCertificate", true)
        config.addDataSourceProperty("trustCertificateKeyStoreUrl", this.keyStoreFile.toURI().toURL())
        config.setJdbcUrl(jdbcUrl)
        config.setUsername(jdbcUsername)
        String password = DataSourceUtils.generateAuthToken(authTokenGenerator, config)
        config.setPassword(password)
        config.setPoolName(tenantId)

        HikariDataSource hikariDataSource = new HikariDataSource(config)

        Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate(
                new IamWorker(tenantId, hikariDataSource, authTokenGenerator),
                THREAD_SCHEDULE_TIME,   // wait x minutes for first execution
                THREAD_SCHEDULE_TIME,   // repeat the task every x minutes
                TimeUnit.MINUTES
        )

        return hikariDataSource
    }
```
**Please note that in the code sample here I have not set maxLifeTime on the config object and it will take on the default value of 30 minutes.**

The IamWorker class passed as a Runnable to the ScheduledExecutorService that refreshes the password dynamically is as follows:
```
/**
     * This runnable is used for generating IAM credentials
     * using the aws rds sdk and setting the password on the
     * supplied data source to the new generated token
     */
    private class IamWorker implements Runnable {

        HikariDataSource hikariDataSource
        RdsIamAuthTokenGenerator tokenGenerator

        IamWorker(HikariDataSource hikariDataSource, RdsIamAuthTokenGenerator tokenGenerator) {
            this.hikariDataSource = hikariDataSource
            this.tokenGenerator = tokenGenerator
        }

        @Override
        void run() {
            LogUtils.logMessage(Level.INFO, "${Thread.currentThread().getId()} Refreshing IAM credentials...")
            String password = DataSourceUtils.generateAuthToken(this.tokenGenerator, this.hikariDataSource)
            this.hikariDataSource.getHikariConfigMXBean().setPassword(password)
        }
    }
```

This code seems to be doing what I expect and the password is refreshed every 14 minutes. However **after 30 minutes** which is the default max life time of a connection it is unable to connect to RDS and all tenant pools lose all their connections as they drop the connection due to the max life time being met and when the connection attempts to get renewed access denied exceptions are thrown. This results in the tenant pools to eventually have no connections shown in the following log:
```
03-Apr-2019 10:53:21.913 [Data Source Thread: default] DEBUG com.zaxxer.hikari.pool.HikariPool - default - Pool stats (total=0, active=0, idle=0, waiting=0)
```
Due to this I started experimenting with the maxLifeTime setting for each tenant data config. **I set the tenant data source maxLifeTime to be 30 seconds**, and I set the IamWorker thread to refresh the RDS credentials every 14 minutes (the original time). I was expecting that for the first 14 minutes any connection closed and renewed would work fine as the password was not dynamically changed and the original RDS token was being used. I was then expecting after the 14 minutes once the credentials were refreshed that the next batch of connections to close and get renewed would fail as the credentials had changed. To my surprise when the credentials were refreshed after 14 minutes this was not the case it worked and no connections failed to authenticate.

Data Source logic with maxLifeTime set to 30 seconds:
```
    /**
     * Creates a data source with the given jdbc url and jdbc username
     * the password is generated for RDS access and the connection uses SSL
     *
     * @param tenantId the id of the tenant which this data source will belong to
     * @param jdbcUrl the url of the data source
     * @param jdbcUsername the username to use to access the data source
     * @return the data source
     */
    private HikariDataSource createDataSource(String tenantId, String jdbcUrl, String jdbcUsername){
        HikariConfig config = new HikariConfig()
        config.setDriverClassName(dataSourceProperties.getDriverClassName())
        config.addDataSourceProperty("useSSL", true)
        config.addDataSourceProperty("requireSSL", true)
        config.addDataSourceProperty("verifyServerCertificate", true)
        config.addDataSourceProperty("trustCertificateKeyStoreUrl", this.keyStoreFile.toURI().toURL())
        config.setJdbcUrl(jdbcUrl)
        config.setUsername(jdbcUsername)
        String password = DataSourceUtils.generateAuthToken(authTokenGenerator, config)
        config.setPassword(password)
        config.setMaxLifetime(30000) // 30 seconds
        config.setPoolName(tenantId)

        HikariDataSource hikariDataSource = new HikariDataSource(config)

        Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate(
                new IamWorker(tenantId, hikariDataSource, authTokenGenerator),
                THREAD_SCHEDULE_TIME,   // wait x minutes for first execution
                THREAD_SCHEDULE_TIME,   // repeat the task every x minutes
                TimeUnit.MINUTES
        )

        return hikariDataSource
    }
```
With this code working, I decided to keep the maxLifeTime setting as 30 seconds and change the IamWorker to run every 14 seconds rather than minutes so that I could confirm that it was definitely working. I left this run for some time and it also worked.

I then decided to put the IamWorker time back to its original 14 minutes as I was happy that credentials were being refreshed correctly. I also changed the maxLifeTime setting higher to **25 minutes** out of curiosity and once again it failed similar to how it failed after the default 30 minutes mark.

I am aware I am using an older version of Hikari but any help as to why setting the maxLifeTime to be short lived as I have shown results in connections to be authenticated fine and when they are longer lived it fails. It also may be the case that I don't fully understand how Hikari deals with getting connections, so any advice or help would be greatly appreciated.

It may help to know I have checked my RDS instance and the max life time for connections is set to 8 hours so I do not think that RDS is terminating the connections.

Thank you.

**Comment by brettwooldridge** [comment]
This seems likely to have more to do with the authentication subsystem than HikariCP.

Once credentials are created, how long are they good for?  Do their lifetimes start at creation time or upon first use?

Turning on debug logging for HikariCP could definitely shed light on lifetime and creation events.

When new credentials are generated, you might try soft-eviction in the pool.  That will close all idle connections and renew them.  In-use connections will be marked and evicted as soon as the return to the pool.  With this behavior, keeping maxLifetime at default is fine, as the lifetime would now be controlled by the credentials renewal thread.

**Comment by deanfundrecs** [comment]
Hi @brettwooldridge thanks for taking the time to respond. 

I spotted that there was a fix made in version 3.2.0 which was brought to your attention in this issue 
[https://github.com/brettwooldridge/HikariCP/issues/1137](url) and was fixed here [https://github.com/brettwooldridge/HikariCP/commit/851e2d4592b52a9c367ada2c76f013b1d4e20ac3](url).

Upgrading to the latest version solved my issue and the code snippet I supplied above worked as is with the new upgrade.

Thank you for the response.
